File: behavioral/dart/visitor_structured/file_system.dart
import 'interfaces.dart';

class File implements FileSystemNode {
  final String _name;
  final int _size;
  
  File(this._name, this._size);
  
  @override
  void accept(FileSystemVisitor visitor) {
    visitor.visitFile(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  int getSize() => _size;
}

class Directory implements FileSystemNode {
  final String _name;
  final List<FileSystemNode> _children = [];
  
  Directory(this._name);
  
  void addChild(FileSystemNode child) {
    _children.add(child);
  }
  
  @override
  void accept(FileSystemVisitor visitor) {
    visitor.visitDirectory(this);
  }
  
  @override
  String getName() => _name;
  
  @override
  int getSize() {
    return _children.fold(0, (total, child) => total + child.getSize());
  }
}

class SizeCalculator implements FileSystemVisitor {
  int _totalSize = 0;
  
  @override
  void visitFile(File file) {
    _totalSize += file.getSize();
    print('File: ${file.getName()} - ${file.getSize()} bytes');
  }
  
  @override
  void visitDirectory(Directory directory) {
    print('Directory: ${directory.getName()} - ${directory.getSize()} bytes');
    _totalSize += directory.getSize();
  }
  
  int get totalSize => _totalSize;
}

class XMLExporter implements FileSystemVisitor {
  @override
  void visitFile(File file) {
    print('  <file name="${file.getName()}" size="${file.getSize()}"/>');
  }
  
  @override
  void visitDirectory(Directory directory) {
    print('  <directory name="${directory.getName()}" size="${directory.getSize()}">');
  }
}

File: behavioral/typescript/visitor_structured/file_system.ts
import { FileSystemVisitor, FileSystemNode } from './interfaces';

export class File implements FileSystemNode {
  private name: string;
  private size: number;
  
  constructor(name: string, size: number) {
    this.name = name;
    this.size = size;
  }
  
  accept(visitor: FileSystemVisitor): void {
    visitor.visitFile(this);
  }
  
  getName(): string { return this.name; }
  getSize(): number { return this.size; }
}

export class Directory implements FileSystemNode {
  private name: string;
  private children: FileSystemNode[] = [];
  
  constructor(name: string) {
    this.name = name;
  }
  
  addChild(child: FileSystemNode): void {
    this.children.push(child);
  }
  
  accept(visitor: FileSystemVisitor): void {
    visitor.visitDirectory(this);
  }
  
  getName(): string { return this.name; }
  getSize(): number {
    return this.children.reduce((total, child) => total + child.getSize(), 0);
  }
}

export class SizeCalculator implements FileSystemVisitor {
  private totalSize: number = 0;
  
  visitFile(file: any): void {
    this.totalSize += file.getSize();
    console.log(`File: ${file.getName()} - ${file.getSize()} bytes`);
  }
  
  visitDirectory(directory: any): void {
    console.log(`Directory: ${directory.getName()} - ${directory.getSize()} bytes`);
    this.totalSize += directory.getSize();
  }
  
  getTotalSize(): number {
    return this.totalSize;
  }
}

export class XMLExporter implements FileSystemVisitor {
  visitFile(file: any): void {
    console.log(`  <file name="${file.getName()}" size="${file.getSize()}"/>`);
  }
  
  visitDirectory(directory: any): void {
    console.log(`  <directory name="${directory.getName()}" size="${directory.getSize()}">`);
  }
}
